OpenRA.Mods.Common/Traits/Sound/AmbientSound.cs

```diff
@@ -22,43 +22,52 @@ class AmbientSoundInfo : UpgradableTraitInfo
 		[Desc("Interval between playing the sound (in ticks).")]
 		public readonly int Interval = 0;
 
-		public override object Create(ActorInitializer init) { return new AmbientSound(init.Self, this); }
+		public override object Create(ActorInitializer init) { return new AmbientSound(this); }
 	}
 
-	class AmbientSound : UpgradableTrait<AmbientSoundInfo>, ITick
+	class AmbientSound : UpgradableTrait<AmbientSoundInfo>, ITick, INotifyRemovedFromWorld
 	{
 		ISound currentSound;
 		bool wasDisabled = true;
 		int interval;
+		WPos cachedPosition;
 
-		public AmbientSound(Actor self, AmbientSoundInfo info)
+		public AmbientSound(AmbientSoundInfo info)
 			: base(info)
 		{
 			interval = info.Interval;
 		}
 
 		public void Tick(Actor self)
 		{
-			if (IsTraitDisabled)
+			if (IsTraitDisabled || !self.IsInWorld)
 			{
-				Game.Sound.StopSound(currentSound);
-				currentSound = null;
-				wasDisabled = true;
+				StopSound();
```

Comment:
Why are you doing this twice? (here and `INotifyRemovedFromWorld.RemovedFromWorld`).  You should be able to remove one of these cases.